File: linked_lists/linked_list.rb
require 'pry'

class Linked_list
  attr_accessor :head

  def initialize node=nil
    @head = node
    @kth = nil
  end

  def last_element
    return head unless head.next
    runner = self.head
    while runner.next
      runner = runner.next
    end
    runner
  end

  def add_node node
    if self.head
      last_element.next = node
    else
      self.head = node
    end
  end

  def add_node_by_value value
    add_node Node.new(value)
  end

  def view_list
    runner = @head
    while runner
      p runner.value
      runner = runner.next
    end
  end

  def length
    count = 0
    runner = head
    while runner
      count += 1
      runner = runner.next
    end
    count
  end

  def remove_node target
    if target == head
      @head = head.next
      return
    end
    previous = head
    runner = head.next
    while runner
      if runner == target
        previous.next = runner.next
        return runner
      end
      previous = runner
      runner = runner.next
    end
    nil
  end

  def contains_node? node
    runner = head
    while runner
      return true if runner == node
      runner = runner.next
    end
    nil
  end

  def contains_value? target
    runner = head
    while runner
      return true if runner.value == target
      runner = runner.next
    end
    nil
  end

  def kth_to_last k
    find_kth head, k
    @kth
  end

  def find_kth head, k
    return 0 unless head
    result = (find_kth head.next, k) + 1
    if result == k
      @kth = head
    end
    return result
  end


  private
end

class Node
  attr_accessor :value, :next

  def initialize value=nil
    @value = value
    @next = nil
  end
end

def partition_linked_list list, target
  small_list = Linked_list.new
  big_list = Linked_list.new
  runner = list.head

  while runner
    next_element = runner.next
    runner.next = nil
    runner.value < target ? (small_list.add_node runner) : (big_list.add_node runner)
    runner = next_element
  end



  small_list.last_element.next =  big_list.head
  small_list
end

def add_two_lists list1, list2
  sum = Linked_list.new
  runner1 = list1.head
  runner2 = list2.head
  remainder = 0


  while runner1 || runner2 || (remainder > 0)
    value_of_column = (runner1.value || 0) + (runner2.value || 0) + remainder
    sum.add_node (Node.new(value_of_column%10))
    remainder = value_of_column/10 > 0 ? 1 : 0
    runner1 = runner1.next
    runner2 = runner2.next
  end

  sum.view_list
end

def add_two_lists_forward_order list1, list2

  arabic1 = (arabic_value_of list1.head).to_i
  arabic2 = (arabic_value_of list2.head).to_i

  sum = arabic1 + arabic2
  arabic_to_forward_linked_list sum
end

def arabic_value_of head
  if head.next
    return head.value.to_s + (arabic_value_of head.next)
  end
  return head.value.to_s
end

def arabic_to_forward_linked_list num
  answer = Linked_list.new
  process_arabic_to_forward_ll answer, num
  answer
end

def process_arabic_to_forward_ll list, num
  if num/10 > 0
    process_arabic_to_forward_ll list, num/10
  end
  list.add_node(Node.new(num%10))
end



# x = Linked_list.new
# list = [1,2,3,4,5,6,7]
# list.each {|val| x.add_node_by_value val }
# head_of_circle = Node.new(8)
# x.add_node head_of_circle
# [9,10,11,12,13].each {|val| x.add_node(Node.new(val))}
# bad_node = Node.new(14)
# x.add_node bad_node
# bad_node.next = head_of_circle


# list = [5,9,2]
# y = Linked_list.new
# list.each {|val| y.add_node_by_value val }
# x.view_list
# p (x.kth_to_last 2).value
# x.view_list
p 'XXXXXXXXXXXXXXXXXXXXXXXX'
# y = partition_linked_list x,5
# y.view_list
# p 'XXXXXXXXXXXXXXXXXXXXXXXX'
# add_two_lists_forward_order x,y
# p 'XXXXXXXXXXXXXXXXXXXXXXXX'
# a = add_two_lists_forward_order x,y
# a.view_list

File: linked_lists/linked_list.js
var Node = function(value){
  timer = Date.now()
  while (Date.now() === timer){}
  this.value = value
  this.id = Date.now()
  this.next = null
};

var LinkedList = function(node){
  this.head = node
  this.next = null
};

LinkedList.prototype.viewList = function(){
  var runner = this.head
  while (runner){
    console.log(runner.value + '  id: ' + runner.id )
    runner = runner.next
  }
}

LinkedList.prototype.lastElement = function(){
  if (this.head){
    var current = this.head
    while (current.next){
      current = current.next
    }
    return current;
  }
  return null
}

LinkedList.prototype.addNode = function(node){
  if (this.head){
    var lastNode = this.lastElement()
    lastNode.next = node
  } else {
    this.head = node
  }
}

LinkedList.prototype.length = function(){
  var count = 0
  var runner = this.head
  while (runner){
    count++
    runner = runner.next
  }
  return count
}

LinkedList.prototype.removeNode = function(node){
  if (node == this.head){
    var temp = this.head
    this.head = this.head.next
    return temp
  }
  var runner = this.head.next
  var previous = this.head

  while (runner){
    if (runner == node){
      previous.next = runner.next
      return runner
    }
    previous = runner
    runner = runner.next
  }
}

LinkedList.prototype.removeDuplicates = function(){
  var duplicatesHash = {}
  runner = this.head
  while (runner){
    if (duplicatesHash[runner.value]){
      this.removeNode(runner)
    } else {
      duplicatesHash[runner.value] = true
    }
    runner = runner.next
  }
}

LinkedList.prototype.removeDuplicatesNoBuffer = function(){
  var current = this.head
  while (current){
    findAndRemoveDups (current, this)
    current = current.next
  }
}

var findAndRemoveDups = function(targetNode, classInstance){
  var runner = targetNode.next
  while (runner){
    if (runner.value == targetNode.value){ classInstance.removeNode(runner)}
    runner = runner.next
  }
}

var partitionLinkedList = function(list, target){
  var smallList = new LinkedList()
  var bigList = new LinkedList()
  var runner = list.head

  while(runner){
    var nextNode = runner.next
    runner.next = null
    if (runner.value < target){
      smallList.addNode(runner)
    } else {
      bigList.addNode(runner)
    }
    runner = nextNode
  }
  smallList.addNode(bigList.head)
  return smallList
}

var linkedListAdder = function(list1, list2){
  arrabic1 = llToArabic(list1.head, 0)
  arrabic2 = llToArabic(list2.head, 0)
  return arabicToLl(arrabic1 + arrabic2)
}

var llToArabic = function(head, powerOfTen){
  if (!head){
    return 0
  }
  return llToArabic(head.next, powerOfTen + 1) + Math.pow(10, powerOfTen) * head.value
}

var arabicToLl = function(num){
  var list = new LinkedList( new Node(num%10))
  num = Math.floor(num/10)
  while (num > 0){
    list.addNode(new Node(num%10))
    num = Math.floor(num/10)
  }
  return list
}

var headOfCircularList = function(list){
  var histo = {};
  var runner = list.head

  while (runner){
    if (histo[runner.id]){
      return runner
    } else {
      histo[runner.id] = true
    }
    runner = runner.next
  }
}

var isPalindrome = function(list){
  var listContents = []
  var runner = list.head
  while (runner){
    listContents.push(runner.value)
    runner = runner.next
  }
  while (listContents.length > 1){
    if (listContents.pop() != listContents.shift()){
      return false
    }
  }
  return true
}





var list = new LinkedList(new Node(1))
var x = [2,3,4,5,5,4,3,2,1]
for (var index = 0; index < x.length; index++){
  list.addNode(new Node(x[index]))
}

console.log(isPalindrome(list))

// var headOfCircle = new Node(8)
// list.addNode(headOfCircle)


// var list2 = new LinkedList(new Node(6))
// var x = [9,10,11,12]
// for (var index = 0; index < x.length; index++){
//   list.addNode(new Node(x[index]))
// }

// var lastElement = new Node(13)
// // lastElement.next = headOfCircle
// list.addNode(lastElement)


// list.viewList()
console.log('XXXXXXXXXXX')


// var hOC = headOfCircularList(list)
// console.log(hOC.value)



// list.removeDuplicatesNoBuffer()
// list.viewList()
// var partitionedList = partitionLinkedList(list,4)
// list.viewList()
// console.log(llToArabic(list.head,0))
// console.log(arabicToLl(7651).head.next.next.next.value)

// var a = linkedListAdder(list, list2)
// console.log(llToArabic(list.head,0))
// console.log(llToArabic(list2.head,0))
// console.log(a.head.next.next.next.value)

